src/Synth/LFO.cpp

```diff
@@ -173,44 +182,41 @@ void LFO::releasekey()
     // and divide by current ramp factor. It will be multiplied during fading out.
     outStartValue *= (1.0f - rampOnRelease);
     // store current time
-    releaseTimestamp = lfopars_.time->time();
+    releaseTimestamp = lfopars.time->time();
     // calculate fade out duration in frames
-    fadeOutDuration = lfopars_.fadeout * lfopars_.time->framesPerSec();
+    fadeOutDuration = lfopars.fadeout * lfopars.time->framesPerSec();
     // set fadeout state
     lfo_state = lfo_state_type::fadingOut;
 }
-
 float LFO::lfoout()
 {
     //update internals
-    if ( ! lfopars_.time || lfopars_.last_update_timestamp == lfopars_.time->time())
+    if ( ! lfopars.time || lfopars.last_update_timestamp == lfopars.time->time())
     {
-        waveShape = lfopars_.PLFOtype;
-        int stretch = lfopars_.Pstretch;
-        if(stretch == 0)
-            stretch = 1;
+        updatePars();
+        switch(lfopars.fel) {
         
-        const float lfostretch = powf(basefreq_ / 440.0f, (stretch - 64.0f) / 63.0f);
-
-        float lfofreq = lfopars_.freq * lfostretch;
-        phaseInc = fabsf(lfofreq) * dt_;
-
-        switch(lfopars_.fel) {
             case consumer_location_type_t::amp:
-                lfointensity = lfopars_.Pintensity / 127.0f; // [0...1]
+                lfointensity = lfopars.Pintensity / 127.0f; // [0...1]
                 break;
             case consumer_location_type_t::filter:
-                lfointensity = lfopars_.Pintensity / 127.0f * 4.0f; // [0...4] octaves
+                lfointensity = lfopars.Pintensity / 127.0f * 4.0f; // [0...4] octaves
                 break; 
             case consumer_location_type_t::freq:
             case consumer_location_type_t::unspecified:
-                lfointensity = powf(2, lfopars_.Pintensity / 127.0f * 11.0f) - 1.0f; // [0...2047] cent
+                lfointensity = powf(2, lfopars.Pintensity / 127.0f * 11.0f) - 1.0f; // [0...2047] cent
                 break;
         }
     }
     
-    float out = baseOut(waveShape, phase);
-    // Apply intensity
+    // refresh freq if tempo has changed
+    if (lfopars.numerator && lfopars.denominator && tempo != time.tempo) {
```

Comment:
Where is freq refreshed if only numerator or denominator change?

Reply:
line 96 in updatePars()

Reply:
So numerator and denominator are part of `lfopars`, so they update `lfopars.last_update_timestamp`, and so are handled by `updatePars` line 196, but tempo is not part of `lfopars`, which means we have to additionally check it here?

Reply:
yes. tempo is in the global used time structure. if it's changed we refresh, but only if bpm sync is active (both numerator and denominator !=0)